### PR TITLE
Unit test using a templates and added features

### PR DESCRIPTION
**Pull request by coax1d**
Sort of forced using templates here for added experience working with templates/specializations.
Added some additional minor features in order to clean things up. Let me know when you are available to discuss this one over a call.
I have a couple questions.

### CODE REVIEW

File: Alien.cpp [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+
 void foreign_life::Alien::set_type(int type) {
     this->type = type;
 }
 
-void foreign_life::Alien::print_type() const {
-    std::cout << "Type of alien is " << this->type
-         << std::endl;
+int foreign_life::Alien::get_type() const {
+    return this->type;
```
Comment (MichaelCook) [inline Alien.cpp]: I think this style of explicitly using `this->` is highly unusual.
The vast majority of C++ coders would just write `type` not `this->type`.

File: unit_test.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#include <unistd.h>
+#include <thread>
+#include <memory>
+#include <mutex>
+#include <iostream>
+#include <assert.h>
+
+#undef NDEBUG
+
+#ifndef NDEBUG
```
Comment (MichaelCook) [inline unit_test.cpp]: Well, you don't want this ifndef in your unit test. You just want the ASSERT macro to always be enabled

File: unit_test.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+#   define ASSERT(condition, message) \
+    do { \
+        if (! (condition)) { \
+            std::cerr << "Assertion `" #condition "` failed in " << __FILE__ \
+                      << " line " << __LINE__ << ": " << message << std::endl; \
+            std::terminate(); \
```
Comment (MichaelCook) [inline unit_test.cpp:16]: Could `throw std::runtime_error();`

Comment (coax1d) [reply unit_test.cpp:16]: Ahh ok if i wanted to instead catch it and move on with other tests then report at the end. I think this is fine for now but that is a nice way of doing it.

File: Alien.cpp (hunk 1 above)
Comment (coax1d) [reply Alien.cpp]: Is this a new feature of modern C++? I just remember in school always being taught to use this pointer

File: BaseballPlayer.h [hunk 4]
```diff
@@ -10,10 +11,14 @@ class BaseballPlayer : public Human {
...
 
-        const std::string& get_name() const;
+        const std::string & get_name() const;
+        void set_name(std::string name);
+
         bool has_baseball() const;
         void catch_baseball();
         void throw_baseball(BaseballPlayer &catch_partner);
+        void put_on_mitt(BaseballMitt mitt);
+        const BaseballMitt & get_mitt();
```
Comment (coax1d) [inline BaseballPlayer.h:22]: This ruins the whole pimpl idiom forward declaration I think right?

**Pull request merged**